jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -1027,54 +1057,68 @@ public void run()
             try
             {
                 Runnable job = null;
-                while (true)
+                exit: while (true)
                 {
                     // If we had a job,
                     if (job != null)
                     {
+                        // signal that we are idle again,
                         idle = true;
-                        // signal that we are idle again
                         if (!addCounts(0, 1))
                             break;
                     }
-                    // else check we are still running
+                    // else check we are still running.
                     else if (_counts.getHi() == Integer.MIN_VALUE)
                     {
                         break;
                     }
 
                     try
                     {
-                        // Look for an immediately available job
-                        job = _jobs.poll();
+                        long idleTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(getIdleTimeout());
+                        job = idleJobPoll(idleTimeoutNanos);
                         if (job == null)
                         {
-                            // No job immediately available maybe we should shrink?
-                            long idleTimeout = getIdleTimeout();
-                            if (idleTimeout > 0 && getThreads() > _minThreads)
+                            // No job available after an idle timeout, maybe we should shrink?
+                            if (idleTimeoutNanos > 0 && getThreads() > getMinThreads())
                             {
-                                long last = _lastShrink.get();
                                 long now = NanoTime.now();
-                                if (NanoTime.millisElapsed(last, now) > idleTimeout && _lastShrink.compareAndSet(last, now))
+                                long shrinkPeriod = idleTimeoutNanos / getIdleTimeoutMaxShrinkCount();
+                                if (LOG.isDebugEnabled())
+                                    LOG.debug("Shrink check, period={}ms {}", TimeUnit.NANOSECONDS.toMillis(shrinkPeriod), QueuedThreadPool.this);
+                                while (true)
                                 {
-                                    if (LOG.isDebugEnabled())
-                                        LOG.debug("shrinking {}", QueuedThreadPool.this);
-                                    break;
+                                    long lastShrink = _lastShrink.get();
+                                    long prevShrink = lastShrink;
+                                    // If the shrink window is too far in the past,
+                                    // advance it to be one idle timeout before now.
+                                    if (NanoTime.elapsed(lastShrink, now) > idleTimeoutNanos)
+                                        prevShrink = now - idleTimeoutNanos;
```

Comment:
Could we just do:
```java
    long shrinkThreshhold = Math.max(lastShrink, now - idleTimeoutNanos);
```
or does nano wrapping make this not work?  It is the semantic we want.

Reply:
doh!